bin/wasm-node/rust/src/json_rpc_service.rs

```diff
@@ -1004,48 +1143,60 @@ impl JsonRpcService {
 
         let client = self.clone();
+        let mut source_disconnected_rx = subscriptions.source_disconnected_rx.clone();
 
         // Spawn a separate task for the subscription.
         (self.tasks_executor.lock().await)(Box::pin(async move {
             // Send back to the user the confirmation of the registration.
-            client.send_back(&confirmation);
+            client.send_back(&confirmation, source_id);
 
             loop {
                 // Wait for either a new block, or for the subscription to be canceled.
                 let next_block = blocks_list.next();
                 futures::pin_mut!(next_block);
-                match future::select(next_block, &mut unsubscribe_rx).await {
-                    future::Either::Left((block, _)) => {
-                        let header = header_conv(header::decode(&block.unwrap()).unwrap());
-                        client.send_back(&smoldot::json_rpc::parse::build_subscription_event(
-                            "chain_newHead",
-                            &subscription,
-                            &serde_json::to_string(&header).unwrap(),
-                        ));
-                    }
-                    future::Either::Right((Ok(unsub_request_id), _)) => {
-                        let response = methods::Response::chain_unsubscribeAllHeads(true)
-                            .to_json_response(&unsub_request_id);
-                        client.send_back(&response);
-                        break;
-                    }
-                    future::Either::Right((Err(_), _)) => break,
+                let inner = future::select(next_block, &mut unsubscribe_rx);
+
+                match future::select(&mut source_disconnected_rx, inner).await {
+                    future::Either::Left(_) => break,
+                    future::Either::Right((inner, _)) => match inner {
+                        future::Either::Left((block, _)) => {
+                            let header = header_conv(header::decode(&block.unwrap()).unwrap());
+                            client.send_back(
+                                &smoldot::json_rpc::parse::build_subscription_event(
+                                    "chain_newHead",
+                                    &subscription,
+                                    &serde_json::to_string(&header).unwrap(),
+                                ),
+                                source_id,
+                            );
+                        }
+                        future::Either::Right((Ok(unsub_request_id), _)) => {
+                            let response = methods::Response::chain_unsubscribeAllHeads(true)
+                                .to_json_response(&unsub_request_id);
+                            client.send_back(&response, source_id);
+                            break;
+                        }
+                        future::Either::Right((Err(_), _)) => break,
+                    },
```

Comment:
While this technically solves the race condition, it only does so in a single-threaded environment and because we don't have use any other `await` after getting notified of a new block. I'd like to veto this solution.

Reply:
😮‍💨 what do you recommend instead?